### DISCUSSION TITLE
instantiating an Array from a GPU memory buffer without GPU mem -> system mem -> GPU mem roundtrip

### DISCUSSION THREAD

**Discussion by hrbigelow** [jax-ml/jax#20506] [Q&A] [answered]
Hi,

I wanted to use jax together with moderngl (an opengl wrapper for python) to build an app that allows the user to visualize and interact with images, but also leverages jax to do various image processing.  moderngl offers the [Buffer](https://moderngl.readthedocs.io/en/5.8.2/reference/buffer.html#buffer) as a handle for GPU memory, but as far as I know it's not possible to instantiate a `jax.Array` directly from a buffer handle to GPU memory from a separate library?  Or is there some lower interoperability layer to achieve this?

I can follow the route of:

1. transfer the `moderngl.Buffer()` data into system memory `bytearray` 
2. Instante with `jnp.array(data: bytearray, ...)`
3. move jax.Array to GPU

But if possible I'd like to avoid the GPU mem -> system mem -> GPU mem roundtrip.

Thanks very much!

**Comment by jakevdp** [comment, marked as answer]
The primary method for zero-copy transfer of GPU buffers into JAX is via the [dlpack](https://dmlc.github.io/dlpack/latest/) data structure, using the [`jax.from_dlpack`](https://jax.readthedocs.io/en/latest/_autosummary/jax.dlpack.from_dlpack.html) function.

If dlpack is not an option, JAX also has support for [`__cuda_array_interface__`](https://numba.pydata.org/numba-doc/dev/cuda/cuda_array_interface.html).

**Reply by hrbigelow** [reply]
Thank you Jake.  This looks to be exactly what I was hoping for.

**Reply by jakevdp** [reply]
I should note that the Python Array API specification landed on `dlpack` as the preferred option for this (see discussion here: https://data-apis.org/array-api/latest/design_topics/data_interchange.html) so if you're implementing something, I'd lean toward that rather than the cuda array interface.

**Reply by hrbigelow** [reply]
Thanks Jake.  Just for completeness, I got a [response](https://github.com/moderngl/moderngl/issues/684#issuecomment-2030142903) from the creator of moderngl and have passed on your comments.  This could be a nice combination for my use case.

**Reply by gszep** [reply]
@hrbigelow you may also be interested in this CuPy / OpenGL interop example :) https://gist.github.com/keckj/e37d312128eac8c5fca790ce1e7fc437

**Reply by mohawk2** [reply]
As I also noted on that moderngl issue, I think this could be solved if the DLPack folks accept this: https://github.com/dmlc/dlpack/issues/183
